### ISSUE TITLE
Request for evaluation/feedback

### ISSUE THREAD

**Issue by daKmoR** [DanielSchaffer/webpack-babel-multi-target-plugin#13]
hey,
we are thinking about using your plugin as part of our [open-wc](https://github.com/open-wc/open-wc)  recommended bundling configuration for webpack.

We are focusing on webcomponents and how to facilitate them. This still requires quite some polyfills.
So having different builds is absolutely vital.

I think we have a pretty good starting point but we are still struggling with the bundle size.

In the spirit of open source I was wondering if you could take a look do a review and give us some hints? Maybe we are running into https://github.com/DanielSchaffer/webpack-babel-multi-target-plugin/issues/11 and it seems there is a workaround? care to share?

This is the related pull request: https://github.com/open-wc/open-wc/pull/122

This is how you could check it out
```
git clone git@github.com:open-wc/open-wc.git
cd open-wc
git checkout feat/webpack
cd packages/building-webpack
npm i
npm run build
npx http-server ./dist/
```

If you open that page in the latest chrome the download is still >50KB even though it's a tiny tiny tiny demo page. I think we should be able to make it smaller - and yes that is our work as we need to find a way to handle all these polyfills with potentially only 2 versions? (as there is only one nomodule trick)?

or do you know any more?

I'm looking forward to your feedback :)

**Comment by DanielSchaffer** [comment]
So! In my cloned repo, I added the [Webpack Bundle Analyzer](https://www.npmjs.com/package/webpack-bundle-analyzer) to help break down what's going into your bundles.

Here's what you're loading:

**main.js - ~12KB**
Everything in here that isn't your code is from `core-js`. I'm not sure there's much to do here, as what gets included here is going to be determined by your usage (e.g. due to `useBuiltIns: 'usage'` and browserslist. The good news is that it gzips down to 4KB.

**Chunk: 2.js - ~8KB** - `@webcomponents/template`

**Chunk: 3.js - ~2KB** - `@webcomponents/webcomponents-platform`

**Chunk: 6.js - ~36KB** - Mix of `lit-element` (~23KB) and more `core-js` (~11.5KB)

**Chunk: 14.ks - ~2KB** - `demo-app.js`

To be honest, I'm not sure there's a _ton_ to do here as far as slimming things down, but here are some things to consider:
- Gzipped, you're at 24KB, which isn't bad at all.
- A bigger issue than size (ha) might be loading 5 files, some of which are extremely small. I'm not sure why Webpack is splitting the chunks the way it is, but it'd be better for it all to be in a single file. 
- The size might seem silly for a tiny demo app, but to me, it's a bit irrelevant for this kind of thing. Obviously there's going to be a bit of size to providing functionality like this, and it'll seem less silly with an app that actually uses all of it.
- Check your dependencies to make sure they're both shipping ES modules to use as source to facilitate tree shaking. Also make sure that you're using the right mainFields to use them - you might need to add `jsnext:main` if any of your deps use that instead of `module` or `es2015`.

**Comment by DanielSchaffer** [comment]
After taking a closer look at what's going on with the demo app, it's a bit clearer why there are so many chunks. I see the dynamic imports in the demo itself, which explains 14.js.

It looks like `@webcomponents/template` and `@webcomponents/webcomponents-platform` are separate because they're added as two separate dynamic imports (`@open-wc/polyfills-loader`, lines 44-45). I expect you could avoid the separate chunks by adding a file that statically imports both, and then just dynamically importing the one. Same goes for the shadow dom polyfills on lines 54-55. 

The return logic in `needsTemplatePolyfill` seems like it might be inverted as well - I'm guessing templating is "broken" if it doesn't correctly clone all the elements that were added, but it's returning true unless they're empty, which is why those are being loaded for Chrome.

As noted in #11, Babel's polyfilling functionality is mainly limited to JS and doesn't help much with DOM APIs, and since this plugin is primarily focused on that, I'm not sure it'll be of as much utility to you in dealing with issues like this. That said, I'd be interested to know what the groupings are of browser support for the DOM APIs you need are. The `nomodules` strategy this plugin exploits effectively limits you to two bundles in order for the magic to work, but if there are more than two "groupings" that are distinctly different enough, I'm wondering if it might make sense for you to use 3-4 more specialized bundles, and basically just use a switcher script instead of `nomodule`. Dropping, for example, all but Safari 12 for a "most modern" bundle might also allow you to cut out more of the `core-js` imports, trimming down the size of the bundle for most users.

**Comment by daKmoR** [comment]
uh these are already very valuable insights :)
thxxx

for the groupings there would probably be 3-4 that make sense but will make serving more complicated and our goal is also to be as simple as possible... e.g. just take the output and put it on a simple webserver and you are done. There are way more advanced servers that are doing differential serving per user client... but that's always tougher to setup.

So using the nomodule trick I think is pretty smart so maybe we could focus on one "ultra modern" package which hopefully serves almost "nothing" (e.g. maybe just bundled and pare modules replaced and that's it)

and the rest should get the es5 module version that shall work down to IE11.

For all the in-betweens we are a little sorry but that's probably the best we can do. imho

@LarsDenBakker what do you think?

**Comment by DanielSchaffer** [comment]
Well - you're going to be limited in who you can exclude by using `type="module"` and `nomodule`. `type="module"` is supported by Safari going back to 10.1, and including iOS 10.3. Edge has support for it too, and I saw there were some cases in your code where it needed a polyfill as well. Trying to do something like that will require some more complicated logic - either the more "advanced" server like you mentioned, which could look at the user-agent and determine which bundle to serve - or you could also use a loader script to do the checking, but that could slow down your load time since it'd block sideloading of app scripts.

**Comment by daKmoR** [comment]
oh boy you are right... `nomodule` trick only get's you that far...

can we maybe combine it all? how much impact do you think would such a loader have? 

index.html
```html
<script type="module">
  if (supports x and y and z) {
    import('ultra-modern.js'); // e.g. zero polyfills only bare modules
    // hmm webpack can not even create something you can just import with a dynamic import?
    // so probably creating script tags - boohoo
  }
  if (supports x) {
    import('modern.js'); // polyfills x+y
  }
  if (no dynamic import) {
    // create script tag 
  }
  ...
</script>

<script nomodule src="old.js"></script>
```

do you know of a good example of such a loader script?

**Comment by DanielSchaffer** [comment]
I know [Modernizr](https://github.com/Modernizr/Modernizr) does this kind of thing, but to be honest, I haven't used it in a while. I suppose it really depends on the requirements - IIRC, Modernizr was nice because it had all the checks built in, but I'm not sure how heavy it is, or how good it is at only include the checks you actually use and such. I'd check it out, but there's always just rolling your own - essentially doing what you're already doing in your polyfill loader, but adapting it to load bundles. The tricky thing here is that I don't think there's an easy way to reference the webpack entry points/bundles from the source code.

If you're going to go this route, it'd probably be cleanest to just write a webpack plugin that generates it, since then you'd have access to all the assets that get generated, and would be able to add assets of your own. It would also let you generate the code to be either inlined on the HTML page or added as an external script to satisfy the folks with restrictive CSPs (which, to be honest, would be a little more complicated, you'd have to do some hacking to make sure the loading order works right). That kind of thing might be a nice sibling plugin to this one, since while technically it _can_ create more than two bundles (by simply adding more targets to the config), it can't do anything smart with them as far as adding them to a template.

**Comment by daKmoR** [comment]
yeah `Modernizr` is probably way too heavy... and the check we need would probably be not even in there 🙈  So yeah I guess "just" extending the polyfill loader would be the way to go.

And I totally agree that this html file needs to be generated by a webpack plugin. Not yet sure how we could align the targets... but probably just provide they same keys and you should be fine...

interesting ideas - thx for all the tips and inspirations 🤗

**Comment by DanielSchaffer** [comment]
Some thoughts and implementation ideas: 

* Webpack already sort of has an internal templating system - it's how it wraps everything in those `__webpack_require__` statements and sets up the chunk asset loading sequencing. Take look in the webpack source for where instances where it uses `runtimeTemplate`, particularly `lib/web/JsonpMainTemplatePlugin.js`. There might be a way to follow a similar pattern for this.
* define a list of feature checks that can be loaded into a map - e.g. `shadow-dom`: `/* javascript code here */` ... ideally, the code for the checks could be in their own .js files so you can unit test them if needed
* options for the plugin would allow you to define the list of checks for each target - for example:
```
new SuperAwesomeDomPolyfillPlugin({
  targets: [
    { key: 'super-modern', conditions: ['shadow-dom', 'template', 'url-search-params'] },
    { key: 'quasi-modern', conditions: ['shadow-dom', 'url-search-params'] },
    { key: 'old-and-broke' }
  ]
})
```
* Find an appropriate Webpack compilation hook to generate the script - it'd have to be after the assets are already generated maybe `additionalAssets`?), since you'll need the full generated filenames (including hashes, etc). Generate your script, which will basically be a series of `if` statements chaining the conditions defined in the options, and adding `<script>` elements to the DOM for the assets for the corresponding bundle. My plugin leaves its targeting info in the chunk data - you can sort of see how I work with it in `babel.multi.target.html.updater.ts`, so you can match up the keys.

The piece that'd be missing here is actually getting the polyfills included in the bundles. As I mentioned in #11, it'd be REALLY cool to have some sort of mechanism that could compare a `browserslist` definition against a manifest of features and give you a list of required polyfills. I think the data is all available through the caniuse.com data set (which is available on GH: https://github.com/fyrd/caniuse). So someone (with more time than I've got, I can barely keep up with this 22-star project, ha!) could write a translation between the `browserslist` API and that data to come up with the list.

That's a fair bit of work of course. I'll be looking at including per-target modules as part of #11, so once that's done you'll at least be able to hard code the polyfill modules you'll need, which should get you started.

Also, the more I think about this, the more it seems like it should be a subfeature of this plugin, basically as a presentation option. Consider: this plugin's primary function is its ability to take a webpack configuration and generate multiple bundles for different browser targets. The most common way to do this is to have two bundles, and the most common way to "present" those bundles are by using the module/nomodule trick - but that's really only an implementation detail. Using switching logic like this would be another way to "present" the bundles that isn't even necessarily specific to DOM features - it could check for whatever the dev wanted. I think it'd make sense for the checks themselves to be provided by the consumer, which would provide the most flexibility and extensibility, and also limit the responsibility of the plugin to maintain that kind of code. This seems like it'd be a great feature for a 3.0 release, especially since I've been mulling changing up the options format to better accommodate target-specific babel settings - I'm just not sure when I'd be able to get to it.

**Comment by daKmoR** [comment]
> Also, the more I think about this, the more it seems like it should be a subfeature of this plugin, basically as a presentation option.

hihi I actually had the exact same thought... all the info is already there (or could be easily added) so it would only be the rendering. 

how about just
```js
new SuperAwesomeDomPolyfillPlugin({
  conditionFunctions: {
     // direction function to inline
     needsShadowDomPolyfill: () => { 
       return hasShadow ? false : true;
     },
     // function to load and inline - could be a later feature
     needsShadowDomPolyfill: require('@foo/needsShadowDomPolyfill'), 
  }
  targets: [
    { key: 'super-modern', conditions: ['needsShadowDomPolyfill', 'template', 'url-search-params'] },
    { key: 'quasi-modern', conditions: ['needsShadowDomPolyfill', 'url-search-params'] },
    { key: 'old-and-broke' }
  ]
})
```

===> output
```html
<html>
  <!-- ... -->
  <script>
    const needsShadowDomPolyfill = () => { 
      return hasShadow ? false : true;
    }
    // all other condition functions - also those required they get inlined
   
   if (needsShadowDomPolyfill && needOtherPolyfill && ... ) {
      // create script tag with src of webpack bundle
   }
   if (...) {
      // other bundle
   }
 </script>
```

and to be honest I would postpone any integration with `caniuse.com` data... first get it to work then make it fanzy... although this feature I would already classify as fanzcy 🙈

**Comment by DanielSchaffer** [comment]
Yeah, it'd be something like that - though the code for the checks, as well as the references to the polyfills themselves would need to be strings so they could be added to the output, and they'd also have to not use ES5+ syntax (unless we also want to run them through babel).

**Comment by LarsDenBakker** [comment]
How much of a delay would we introduce by feature detecting in JS before loading pur app. Ive played around with it before. We could preload/prefetch the main app in the meantime while we do our polyfill work.

**Comment by DanielSchaffer** [comment]
@LarsDenBakker I'm not sure, as this is relatively new territory for me.

The problem here is that things get more complicated when we're talking about more than two bundles, or targeting browsers by something other than support for ES modules.

With only two bundles targeted by ES module support, this sort of feature could work by hooking in to webpack's existing deferred loading logic - this is how it waits until vendor chunks are loaded before invoking entry point code. The polyfill loader could be injected into the main bundle, it would determine which polyfill bundle is required, and webpack's deferred loading would wait until that polyfill bundle is loaded before starting the app. This results in only a single extra HTTP request, and so long as the polyfill loader code is injected at the beginning of the initial bundle, it's possible that the polyfill could finish loading around the same time as the vendor bundle.

Unfortunately, all that sort of goes out the window with more than two bundles, or using different targeting. Loading any other part of the app would be blocked by loading and executing the polyfill loader, since we'd need to do ES feature checks in the loader to determine even which app and vendor bundles to load. If you don't have a CSP, or have one that allows inline code, it'll be faster, but still not as fast as having the script tags already on the page.

There might be an option for a hybrid approach - use ES module support (or lack thereof) to weed out legacy browsers - those would always get ES5 bundles with all the ES feature polyfills. Browsers with ES module support could get app/vendor bundles with a minimal set of polyfills (or none at all) and use the polyfill loader to sideload both ES and DOM feature polyfills as described above.

**Comment by DanielSchaffer** [comment]
@daKmoR @LarsDenBakker looks like your PR has been merged, so I'm going to close this. If there are specific features you need from this project, absolutely feel free to open a feature request issue.
